dialogue-apache-hc5-client/src/test/java/com/palantir/dialogue/hc5/ErrorRecoveryTest.java

```diff
@@ -0,0 +1,122 @@
+import com.palantir.dialogue.TestEndpoint;
+import io.undertow.Undertow;
+import io.undertow.server.handlers.ResponseCodeHandler;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.util.Optional;
+import javax.net.ssl.SSLContext;
+import org.junit.jupiter.api.Test;
+
+@SuppressWarnings("checkstyle:NestedTryDepth")
+public final class ErrorRecoveryTest {
+
+    @Test
+    public void errorRecovery() throws Exception {
+        Clients clients = DefaultConjureRuntime.builder().build().clients();
+        Undertow server = startServer();
+        try {
+            String uri = "https://localhost:" + getPort(server);
+            ClientConfiguration conf = TestConfigurations.create(uri);
+            try (ApacheHttpClientChannels.CloseableClient client =
+                    ApacheHttpClientChannels.createCloseableHttpClient(conf, "client")) {
+                Channel channel = ApacheHttpClientChannels.createSingleUri(uri, client);
+                EndpointChannel endpointChannel = request -> channel.execute(TestEndpoint.POST, request);
+                assertThatThrownBy(() ->
+                                singleRequest(clients, endpointChannel, Optional.of(ErrorThrowingRequestBody.INSTANCE)))
+                        .isInstanceOf(StackOverflowError.class);
+                // Following a failure, the client should be operable.
+                singleRequest(clients, endpointChannel, Optional.empty());
+            }
+        } finally {
+            server.stop();
+        }
+    }
+
+    private static void singleRequest(Clients clients, EndpointChannel endpointChannel, Optional<RequestBody> body) {
+        clients.callBlocking(endpointChannel, Request.builder().body(body).build(), VoidDeserializer.INSTANCE);
+    }
+
+    private static Undertow startServer() {
+        SSLContext sslContext = SslSocketFactories.createSslContext(TestConfigurations.SSL_CONFIG);
+        Undertow server = Undertow.builder()
+                .setHandler(ResponseCodeHandler.HANDLE_200)
+                .addHttpsListener(0, null, sslContext)
+                .build();
+        server.start();
+        return server;
+    }
+
+    private static int getPort(Undertow undertow) {
+        return ((InetSocketAddress)
+                        Iterables.getOnlyElement(undertow.getListenerInfo()).getAddress())
+                .getPort();
+    }
+
+    private enum VoidDeserializer implements Deserializer<Void> {
+        INSTANCE;
+
+        @Override
+        public Void deserialize(Response response) {
+            response.close();
+            return null;
+        }
+
+        @Override
+        public Optional<String> accepts() {
+            return Optional.empty();
+        }
+    }
+
+    private enum ErrorThrowingRequestBody implements RequestBody {
+        INSTANCE;
+
+        @Override
+        public void writeTo(OutputStream _output) {
+            throw new StackOverflowError();
```

Comment:
This makes for a nice test, and I suspect is similar to the issue which prompted this change, but in practice I don't think httpclient should actually close the connection pool when an error is thrown from this portion of the request. I will follow up separately on that.